### PR TITLE
Try state in context with useHook

### PR DESCRIPTION
**Pull request by david-szabo97**
Resolves: https://github.com/reakit/reakit/issues/745

Alternative implementation to https://github.com/reakit/reakit/pull/792, I felt like #792 more of a hack rather than a solution. That's why I moved on and tried to come up with a solution that doesn't change the core and more reusable.

## Benefits
* Less code, no need to spread those states everywhere
* Better performance, less rerenders

## Goals
* 100% backward compatibility. It should work with every existing component without any changes. This isn't proved yet but I'll include tests later on.
* Should be easy to enable this new feature for existing and upcoming components

## Is there anything we need to do to enable this feature for components?
**Yes**

The PR introduces two new hooks: `useStateContextConsumer` and `useStateContextProvider`. To create a context you can use the helper function `createStateContext`.

#### `useStateContextProvider`
This hook should be used by the parent component. For example `ComboboxPopover` is a component that contains `ComboboxOption`. In this case, `ComboboxPopover` is the parent component and it should provide the context.
  
First, you will need a context: 

```js
export const StateContext = createStateContext<
  Partial<unstable_ComboboxState>
>();
```

Then you need to add the hook to the `compose` array as the very first value:
```js
compose: [useStateContextProvider(StateContext), useComboboxList, usePopover],
```

If your component is using `useComposeProps` too, then you need to add the hook there as well:
```js
useComposeProps(options, { tabIndex, ...htmlProps }) {
  htmlProps = useStateContextProvider(StateContext)(options, htmlProps, true);
  htmlProps = useComboboxList(options, htmlProps, true);
  htmlProps = usePopover(options, htmlProps, true);

  return {
    ...htmlProps,
    tabIndex: tabIndex ?? undefined,
  };
}
```

#### `useStateContextConsumer`
This hook should be used by the child component. For example `ComboboxPopover` is a component that contains `ComboboxOption`. In this case, `ComboboxOption` is the child component and it should consume the context.

You should create a context in your parent component as mentioned above. Then you need to add the hook to the `compose` array as the very first value:
```js
compose: [
  useStateContextConsumer({
    context: StateContext,
    shouldUpdate: (id, state, nextState) => {
      return (
        state.currentId === null ||
        nextState.currentId === null ||
        id === state.currentId ||
        id === nextState.currentId
      );
    },
    updateDependencies: (state) => [state?.currentId],
  }),
  useComboboxItem,
  useCompositeItem,
]
```
`useStateContextConsumer` takes an object with 3 properties:
- `context` the context that you would like to consume, this should be the context that your parent component provides
- `shouldUpdate` a function that returns whether the component should update its state or not. This function gives you 3 parameters: `id` of the component, `state` current state of the component and `nextState` which is the state that we just received from the context. If we return `true`, then the `nextState` will be applied to the component.
- `updateDependencies` under the hood, the pubsub system takes some dependencies. According to these dependencies, your component resubscribes to the pubsub system. This makes sure you receive correct values in `shouldUpdate`.

## How to test?

- `npm run storybook`
- Open `Combobox -> Combobox Non Context Vs Context`

## Does this PR introduce breaking changes?

Hopefully not!

### CODE REVIEW

**Comment by david-szabo97** [comment]
@diegohaz This is an alternative implementation that is not hacking into the `createComponent`. I would like to avoid that to make it more reusable and robust. This way we can totally decouple the `StateContext` and all the related components.

And it is using `createHook` so it's more `Reakit`-ish way.

File: packages/reakit-system/src/useStateContextConsumer.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                  state.currentId === null ||
+                  id === state.currentId ||
+                  id === nextState.currentId
+                ) {
+                  setState(nextState);
+                }
+              })
+            : undefined,
+        [subscribe, state?.currentId, id, htmlProps.id]
+      );
```
Comment (david-szabo97) [inline packages/reakit-system/src/useStateContextConsumer.tsx]: Right now this check is hardcoded here. But we could provide a `function` to check if it should update the state or not. More like how `propsAreEqual` works. So we can make it possible to override this behavior.

**Comment by david-szabo97** [comment]
@diegohaz  @tom-sherman Just finished this up 😄 Tests are passing. I'll experiment with other components as well. Would like to hear your thoughts about this implementation.

**Comment by tom-sherman** [comment]
Amazing work @david-szabo97 🚀 

Just FYI I currently don't have time to review Reakit PRs (I work on react native right now and am currently trying to fight off burnout in these trying times, so avoiding out of hours projects) - I've removed myself from the review request - hope you don't mind!

**Comment by brianespinosa** [comment]
This looks pretty slick, and definitely way better for userland in general. I will give the tires a kick on this on our functional design system sometime next week to see how it feels.

**Comment by diegohaz** [comment]
@brianespinosa Thanks! That will help a lot. I'm more focused on v2 right now, but it would be awesome to have this feature on v1 as well. David has done an excellent job.

**Comment by brianespinosa** [comment]
@diegohaz do we have an alpha or nightly build version for v2 that we can take a look at yet? I did not realize that was coming.

**Comment by diegohaz** [comment]
> @diegohaz do we have an alpha or nightly build version for v2 that we can take a look at yet? I did not realize that was coming.

Not yet! I'm still experimenting with some different implementations to see if/how they work in real projects, especially considering the new concurrent rendering features on React. Once I have something functional I'll create a branch here and propose the changes. I expect this will take a few months, but I hope to have an alpha version or something by the end of the year.

**Pull request closed without merging**